### PR TITLE
feat: introduce automation for stable branches

### PR DESCRIPTION
**Pull request by mert-kirpici**
This commit introduces a github workflow for automating updating the requirements for each of the openstack releases, based on their release page metadata taken from the opendev.org/openstack/releases repository.

To introduce a new release:
- a stable/<release> branch needs to be created
- the update-tempest-releases github workflow needs to be modified

In addition to that, starting from this commit, PR workflow will run against the changes to the stable/* branches as well as master and main.

### CODE REVIEW

**Comment by mert-kirpici** [comment]
Before merging we need to:
- create `stable/<release>` branches
- register `<release>/edge` or `<release>/stable` channels to the snap store
- set up snap packages on launchpad for each stable branch

File: tools/update-snapcraft
Comment (aieri) [inline tools/update-snapcraft:3]: the snap as a whole is licensed under the Apache v2.0 license. Would you explicitly want this script to *not* be under that license?

File: tools/update-snapcraft
Comment (aieri) [inline tools/update-snapcraft:1]: since this script does not have a `.py` extension, it is not picked up by pylint. Running pylint manually yields several recommendations. Can you please add linting this file to `make lint` and - in general - the CI?

File: tools/update-snapcraft
Comment (aieri) [inline tools/update-snapcraft:91]: why import here?

File: .github/workflows/update-snapcraft.yaml
Comment (aieri) [inline .github/workflows/update-snapcraft.yaml:17]: these are the requirements for your update-snapcraft script. Wouldn't it be easier to hold them in a `requirements.txt` file for ease of debugging (i.e. when you may want to step through `update-snapcraft` from your local machine)?

File: tools/update-snapcraft
Comment (aieri) [inline tools/update-snapcraft:36]: why not `max(...)`?

File: tools/update-snapcraft
Comment (aieri) [inline tools/update-snapcraft:77]: [optional] You could consider using the unpacking operator here:
```suggestion
    snapcraft_yaml['parts']['tempest']['python-packages'] = [
        'confluent-kafka==1.8.2',
        *get_latest_plugin_requirements(args.release),
        get_latest_tempestconf_requirements(),
    ]
```

**Review by aieri — CHANGES_REQUESTED** [review]
Overall +1 on the approach; I have a few comments though

File: tools/update-snapcraft
Comment (mert-kirpici) [reply tools/update-snapcraft:1]: done in the update.

File: tools/update-snapcraft
Comment (mert-kirpici) [reply tools/update-snapcraft]: because i hadn't thought of it :), implemented in the update. thanks.

File: .github/workflows/update-snapcraft.yaml
Comment (mert-kirpici) [reply .github/workflows/update-snapcraft.yaml]: introduced `requirements-dev.txt` in the update.

File: tools/update-snapcraft
Comment (mert-kirpici) [reply tools/update-snapcraft]: since `sys` was only needed if the file is run, not if it is imported, i had done it that way. added to the top of the file in the update.

**Review by aieri — COMMENTED** [review]
+1 on the changes but `make lint` fails. I don't understand why rerunning the CI jobs doesn't show the error. Perhaps they don't pick up the latest revision if you simply rerun them?

**Comment by mert-kirpici** [comment]
> +1 on the changes but `make lint` fails. I don't understand why rerunning the CI jobs doesn't show the error. Perhaps they don't pick up the latest revision if you simply rerun them?

Could you share the error you are facing and the git log please? They seem to work fine for me locally and I believe the CI is picking up the changes since the `black` output shows there are 3 files that were processed, this number matches with the changes on the tip of this branch. It used to be 2, in the older workflow runs.

**Review by aieri — APPROVED** [review]
+1, modulo the steps that you mentioned should be performed before this PR can be merged

**Review by esunar — DISMISSED** [review]
Pretty impressive, I am looking forward to see it working.

File: .github/workflows/update-tempest-releases.yaml
Comment (agileshaw) [inline .github/workflows/update-tempest-releases.yaml:35]: Instead of creating one job per release, I'm think if we let `update-snapcraft.yaml` accept a list of openstack releases instead of a single one and use [jobs.<job_id>.strategy.matrix](https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix) to run the update for each release, we could achieve the same goal with a single job here:
```
jobs:
  update-releases:
    uses: ./.github/workflows/update-snapcraft.yaml
    with:
      openstack-releases: "['antelope', 'zed', 'yoga', 'xena', 'wallaby', 'victoria', 'ussuri']"
```
This way we avoid repeating similar codes, and also if we need to add a release in the future, we can simply extend the list

**Review by agileshaw — CHANGES_REQUESTED** [review]
This is a very interesting solution. Overall LGTM, I just have one suggestion for simplifying the implementation.

File: .github/workflows/update-tempest-releases.yaml
Comment (mert-kirpici) [reply .github/workflows/update-tempest-releases.yaml]: hey, thanks for this! indeed it looks like a much better solution, i should've read the workflow docs better :sweat_smile:  

Implemented it and pushed an update. will put up a comment with links to my trial run in my fork. cheers.

**Comment by mert-kirpici** [comment]
Hi @agileshaw @aieri @esunar. I have pushed two small updates and would like a final re-review before we land this.

1 - I discovered a small bug in my implementation during review and mitigated that. The logic that decides to create a pull request using `git diff --exit-code` took all the modified files into account rather than just `snap/snapcraft.yaml`. In the create pr step, we only stage and commit the `snap/snapcraft.yaml` file anyway but this will ensure that the step does not even get run if there are no changes to that specific file we are interested in. The restored `update_snapcraft.yaml` from the default branch might potentially differ from what we have in `stable/` therefore `git diff --exit-code` would report changes even if there are no changes to the `snapcraft.yaml` file. Now we are running `git diff --exit-code snap/snapcraft.yaml` just to be sure.

2 - I applied @agileshaw's suggestion to implement a matrix strategy. Which turned out to be much more concise. I also tested this in my fork an would like you walk you through the steps of what I did and provide some links so that you can take a look at how this works in action.

### Steps
  - pushed the updates to my `main`
  - created `stable/<release>` branches from the tip of `main`
  - run the workflow

[here](https://github.com/mert-kirpici/snap-tempest/actions/runs/6021310048) is a link to the workflow run in my fork.  

### Observations:

- the Create Pull Request step of the antelope job does not even get run, since the changes in this PR includes the antelope revisions of the components. so 6/7 PRs created.
- 6 PRs created in the repo, all look like they have the necessary revisions.

File: .github/workflows/update-snapcraft.yaml
Comment (agileshaw) [inline .github/workflows/update-snapcraft.yaml:39]: nit: Can we have a better name for the pull request? E.g. "update dependencies in snapcraft"

File: .github/workflows/update-snapcraft.yaml
Comment (mert-kirpici) [reply .github/workflows/update-snapcraft.yaml:17]: planning to impement the tox target idea. and thinking hardcoding the dependencies right inside the `tox.ini` rather than a separate requirements file. I don't like the idea of maintaining short files here and there. objections?

Comment (rgildein) [reply .github/workflows/update-snapcraft.yaml:17]: I was also thinking to use only dependencies in `tox.ini`, so :+1:  from me.

File: tools/update_snapcraft.py
Comment (mert-kirpici) [reply tools/update_snapcraft.py:1]: 0 - yeah i faced that error also during testing, however since during CI, the script is only run once I left it as is since I thought running it manually a couple times was not a use case, but it should be easy to implement. will implement it in the update
1 - yeah i think its a good idea to have a tox target for this. I think it should be simple to include it here. will implement in the update
2 - python-tempestconf does not really have any official alignment with o7k releases that I can tell. however it has somewhat short list of requirements. i see no reason to always include the latest release.
3 - python-tempestconf is not included in the o7k release page and the releases repository. thats why we are checking the latest release from pypi.

File: .github/workflows/update-snapcraft.yaml
Comment (mert-kirpici) [reply .github/workflows/update-snapcraft.yaml]: implemented in update

File: .github/workflows/update-snapcraft.yaml
Comment (mert-kirpici) [reply .github/workflows/update-snapcraft.yaml]: to keep the title, after discussion with @rgildein changed the commit message to `release(antelope): update snapcraft.yaml` in the update.

File: tools/update_snapcraft.py
Comment (mert-kirpici) [reply tools/update_snapcraft.py:1]: implemented 0 and 1 in the update.

**Comment by mert-kirpici** [comment]
@aieri @agileshaw @rgildein @esunar 

After #21 got merged, I needed to rebase due to conflict. The workflows apparently won't get run if there is a conflict with the base branch.

The update includes:
- renaming the automation branches from `automation/<release>` to `release/<release>`
- introducing a tox target called `release`
- enabling consequent runs of the `release` target for local usage

File: tools/update_snapcraft.py
Comment (agileshaw) [inline tools/update_snapcraft.py:89]:
```suggestion
    If fast forwarding is not possible, e.g. repo is dirty,
    remove the repo and reclone.
```

File: tools/update_snapcraft.py
Comment (agileshaw) [inline tools/update_snapcraft.py:128]: Personally I think this function is a bit overkill. Why can't we simply remove the cloned directory upon completion of this python script? Then we can have sequential runs, and if the path already exist when cloning the directory (meaning that there's something unknown occupying the directory and we probably shouldn't blindly overwrite it) we can just exit in error.

Adding unnecessary code increases future maintenance burden, so I prefer to keep the logic as simple as possible.

Comment (mert-kirpici) [reply tools/update_snapcraft.py:128]: That was my first instinct as well since i also did not want to write logic for the repo updates. However upon checking the releases repository, I realized its size is 47 MB as of today. When combined with the fact that [pygit2 does not have a `depth` parameter in `clone_repository()`](https://github.com/libgit2/pygit2/issues/630),  

we are left with:
- this approach, which I don't personally find unnecessary or overkill given the circumstances
- subprocess calls to the git cli to clone with depth 1(still comes to 25 MB)
- giving up on local usage feature of the script altogether
- pulling 47 MB with each invocation
- other solutions that don't come to my mind

I am personally at equal distance to any of the above. So I leave it to the maintainers to decide at this point. Will happily update the PR with suggestions.

File: .github/workflows/update-snapcraft.yaml
Comment (mert-kirpici) [reply .github/workflows/update-snapcraft.yaml:29]: alright, will remove in the update. thanks!

File: tools/update_snapcraft.py
Comment (agileshaw) [reply tools/update_snapcraft.py:128]: I see, thanks for the explanation. 

Aside from having complex logic, the part of this solution I worry about the most is blindly overwriting the path if it already exists and it's a git directory (we don't know the actual contents and it could be tracking other remotes). Though I know the chances are low and it only happens during local usage, I feel like we should avoid this risk. If we end up going with this approach, we should at least check the remote origin of the repo and see if it matches RELEASES_REPO_URL before trying to do at operations with it.

A second opinion is appreciated.

Comment (mert-kirpici) [reply tools/update_snapcraft.py:128]: yeah, I agree. that check must be here. if the decision is to keep this solution, I will definitely add that check. thanks for that!  
maybe @esunar or @aieri could give a second opinion about how to move forward about this particular bit here.

Comment (aieri) [reply tools/update_snapcraft.py:128]: tbh I don't expect local usage to be a frequent use case. For this reason, I'd even be ok with the very simple old behavior.
The releases repo being 47 MB is not great though, and not having `--depth` in pygit2 makes it even worse. On the other hand, this script will most of the time be run within the CI.

My proposal would be:
* use `tempfile` instead of `/tmp/releases`
* clean up when the script is done, unless a CLI option is passed to prevent it
* subscribe to https://github.com/libgit2/pygit2/issues/630 and update this code when a depth parameter is added

Comment (rgildein) [reply tools/update_snapcraft.py:128]: Since we are defining the tox testenv, the git clone command can be used before calling this script. It's just an idea, not saying we must do it. Since most time this will be run only with CI it's ok.
One more think, it will be faster since CI will not install any more Python dependencies.

File: tox.ini
Comment (mert-kirpici) [reply tox.ini:72]: the bits that would go here would be `git {clone, checkout, add, commit}`. I can't see the use for automating these for the local user really given there might be a lot of things going wrong in the local repository.  
That being said, as a maintainer, if you require this to be implemented that way, I can go ahead and update the pull request.

File: tools/update_snapcraft.py
Comment (mert-kirpici) [reply tools/update_snapcraft.py]: for simplicity, in the update I added a flag called `--reuse` and added `tempfile` usage, which I have missed in the first iteration.

**Review by agileshaw — COMMENTED** [review]
Overall looks good. I just have one suggestion.

File: pyproject.toml
Comment (aieri) [reply pyproject.toml:39]: allowing pygit2 specifically sounds like a good idea

Comment (mert-kirpici) [reply pyproject.toml:39]: fwiw i had decided to go this way because I didn't like the fact that our requirements are leaking into the configuration.  will change in the update

File: tools/update_snapcraft.py
Comment (rgildein) [inline tools/update_snapcraft.py:125]: Can we add this to some constant, so it will be easier to change it.

File: tox.ini
Comment (rgildein) [reply tox.ini:72]: I think right now it's ok, but I think that it may be useful. We will see. :+1:

File: .github/workflows/update-snapcraft.yaml
Comment (aieri) [reply .github/workflows/update-snapcraft.yaml:11]: Setting `ubuntu-22.04` makes the workflow behavior a bit more predictable; on the other hand the script is simple enough that it should be safe to use `ubuntu-latest`

**Review by aieri — COMMENTED** [review]
Overall +1. I'll wait for the pending conversations to be resolved before giving a formal approval.

File: .github/workflows/update-snapcraft.yaml
Comment (rgildein) [reply .github/workflows/update-snapcraft.yaml:11]: I think it's more useful to use `ubuntu-latest` and if nest year we see CI failure, we can report an issue and handle it.

**Review by agileshaw — COMMENTED** [review]
Overall LGTM. Also waiting for all conversations to be resolved before hitting "Approve".

File: .github/workflows/update-snapcraft.yaml
Comment (mert-kirpici) [reply .github/workflows/update-snapcraft.yaml:11]: ok will change in the update

File: tools/update_snapcraft.py
Comment (mert-kirpici) [reply tools/update_snapcraft.py:124]: do you mean the version or the complete string? can you give an example of what would you like to see here?

Comment (rgildein) [reply tools/update_snapcraft.py:124]: something like this
```python
ADDITIONAL_REQUIREMENTS = [
    "confluent-kafka==1.8.2",
]
```
or if there will no more requirements like this, than use simply
```python
CONFLUENT_KAFKA_VERSION="1.8.2"
```

Comment (mert-kirpici) [reply tools/update_snapcraft.py:124]: ok will update.

**Review by aieri — APPROVED** [review]

**Review by esunar — APPROVED** [review]

**Review by rgildein — APPROVED** [review]
LGTM

**Review by agileshaw — APPROVED** [review]

**Pull request merged**
